Judges - extend invoke_litellm to support tool calling

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/17445?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17445/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17445/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17445/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Judges - extend invoke_litellm to support tool calling

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use default=str for JSON serialization in tool result handling

- Add default=str parameter to json.dumps() for tool results
- Handles non-JSON-serializable objects by converting them to strings
- Consistent with other JSON serialization patterns in the codebase
- Prevents serialization errors for complex tool return types

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor utils.py for better code organization

- Add type hints for Dict and Any imports
- Extract response format to _get_judge_response_format() function
- Remove unnecessary spacing between variable declarations
- Maintain all existing functionality with cleaner code structure

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Clean up judge utils documentation and code organization

- Remove references to 'judge model' in favor of just 'judge'
- Move _get_judge_response_format function below _invoke_litellm
- Remove debug logging for tool registration and invocation
- Add _create_tool_response_message utility for consistent message formatting
- Simplify code by using utility function for tool response messages

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move _create_tool_response_message below _invoke_litellm and fix indentation

- Move utility function below _invoke_litellm for better code organization
- Fix indentation issue in tool call check
- Functions now ordered: _invoke_litellm, _create_tool_response_message, _get_judge_response_format, _get_litellm_retry_policy

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add _create_mlflow_tool_call utility for cleaner code

- Add _create_mlflow_tool_call function to convert LiteLLM tool calls to MLflow ToolCall
- Simplify code by using utility function instead of inline object construction
- Keep imports inside _invoke_litellm to avoid circular import issues
- All tests pass with the refactored code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add test for trace requiring LiteLLM and improve code clarity

- Add test_invoke_judge_model_with_trace_requires_litellm for error path
- Move safe MLflow imports to module level (avoiding circular imports)
- Keep tool imports at function level to avoid circular dependencies
- Use keyword arguments for function calls for better readability
- Update _create_mlflow_tool_call return type hint to ToolCall

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Rename tool-related helpers to include 'litellm' in names

- Rename _create_mlflow_tool_call to _create_mlflow_tool_call_from_litellm
- Rename _create_tool_response_message to _create_litellm_tool_response_message
- Makes it clearer these helpers are specific to LiteLLM integration
- Update test to use kwargs for assertion matching new calling pattern

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Improve type hints in judge utils helper functions

- Remove unnecessary Any type for litellm_tool_call parameter
- Update _create_litellm_tool_response_message return type to dict[str, str]
- Document litellm_tool_call as ChatCompletionMessageToolCall in docstring

Signed-off-by: dbczumar <corey.zumar@databricks.com>

BenWilson2 reviewed (approved):
Nits and a non-blocking execution question. Implementation looks solid and the loop iterator for chained tool calling responses looks great!

dbczumar committed:
Address PR review comments from BenWilson2

- Add comment explaining function-level imports to avoid circular dependencies
- Add TODO comment for potential concurrent tool execution optimization
- Refactor test utility functions into pytest fixtures for cleaner test code
- Keep helper functions for tests that need custom values
- Note: trace parameter already documented in Args section
- Note: No Raises block existed to remove

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/judges/utils.py:
```diff
@@ -97,18 +118,183 @@ def _is_litellm_available() -> bool:
+    Args:
+        provider: The provider name (e.g., 'openai', 'anthropic').
+        model_name: The model name.
+        prompt: The prompt to send to the model.
```
Added! :D

dbczumar committed:
Add trace parameter to _invoke_litellm docstring and remove Raises block

- Add missing trace parameter documentation in _invoke_litellm Args section
- Remove Raises block from _invoke_litellm docstring per PR review feedback

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/judges/utils.py:
```diff
@@ -97,18 +118,183 @@ def _is_litellm_available() -> bool:
+                return message.content
+
+            messages.append(message.model_dump())
+            for tool_call in message.tool_calls:
```
We do, but let's do that as a followup. Added a TODO

dbczumar commented on mlflow/genai/judges/utils.py:
```diff
@@ -97,18 +118,183 @@ def _is_litellm_available() -> bool:
+        TimeoutErrorRetries=num_retries,
+        RateLimitErrorRetries=num_retries,
+        InternalServerErrorRetries=num_retries,
+        ContentPolicyViolationErrorRetries=num_retries,
```
Guardrails should be resilient to a variety of LLM outputs. If an LLM response violates a guardrail, it's a better UX to retry optimistically in hopes of getting a guardrail-compliant response rather than fail.

BenWilson2 replied:
Sounds good to me

dbczumar committed:
Remove duplicate helper functions now that fixtures are available

- Remove create_mock_response, create_mock_tool_response, create_test_trace helper functions
- Remove create_mock_tool helper function
- Inline the mock creation directly in tests that need custom values
- Keep fixtures for standard test cases, inline for custom cases

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix test assertion to match updated error message

- Update test_custom_prompt_judge_llm_error assertion
- Error message changed from 'judge model' to 'judge' in earlier refactor

Signed-off-by: dbczumar <corey.zumar@databricks.com>

BenWilson2 reviewed (approved):
Looks great! 
I just merged master into our feature branch so that we can have the consistent updated clint and ruff rules applied. 
You'll want to `pip install -U  -r requirements/lint-requirements.txt` before running pre-commit locally

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented:
Test failures are unrelated to changes in this PR. Merging into feature branch.
